learningequality/studio
# Markdown editor math rendering fixes

## Description

This change includes a number of usability and styling improvements to the markdown math editor, and adds a vue-based system for implementing custom markdown editor nodes as web components.

#### Issue Addressed

Fixes #2179 
With substantial progress toward fixing #2077

### After
![Peek 2020-09-17 15-39](https://user-images.githubusercontent.com/389782/93525920-4fa6cd00-f8fc-11ea-8e46-41f9ed55141c.gif)

## Steps to Test

Make sure you run `yarn install`, as this includes a new dependency!

Play around with math formulas and see if you can break them.
- [x] Add multiple formulas
- [x] Try deleting the little added buffer spaces before and after the formulas
- [x] Try adding and deleting lots of stuff unrelated to the formulas (this was breaking before)
- [x] Try this on Firefox, Chrome and if you can, Edge (I haven't tested Edge)
- [x] Try somehow selecting and deleting stuff inside of the formulas

## Implementation Notes

#### At a high level, how did you implement this?
I wrapped the formula rendering in a custom element that's hidden by the shadow DOM, and added some extra key handler logic to ensure squire behaves well and consistently across browsers.

There were four main challenges:
1. TUI.editor's underlying library, Squire.js, was reforming the editor contents on most every keypress, but especially on 'backspace' presses.  This resulted in extra added `div` elements getting sprinkled into the MathQuill markup.
- This was fixed by using the shadow DOM to hide the MathQuill elements from squire.js.
- Instead of using the shadow DOM directly, I used a library called `vue-custom-elements`, which wraps Vue components so that they can be used as custom elements.
2. TUI.editor's CSS was interfering with MathQuill's CSS
- At first I fixed this with `!important` directives, but the shadow DOM solution also helped keep MathQuill styles separate.
3. Squire's text editing selection behavior was different on Firefox and Chrome, and wasn't built for custom elements or elements, resulting in the cursor getting stuck or even stuff getting deleted on the wrong side of the cursor.
- Some relatively complex editor selection fixes are applied on every keyDown event.
4. Cursor selection with the mouse only works if custom editor nodes are wrapped in non-breaking spaces, which often get deleted during editing.
- I added some logic to ensure that custom editor nodes will always be wrapped in non-breaking spaces, even if the user happens to delete them.

#### Does this introduce any tech-debt items?
1. The tricks used to correct squire.js selection quirks are pretty tricky, difficult to test and would be much better fixed by modifying squire.js itself.
- I documented them as best as I could.
2. Although custom elements can be polyfilled, the shadow DOM workaround can't be, so if we ever wanted to port the editor to Kolibri we'd have to either fix the underlying issues with TUI.editor/squire.js or potentially swap out the core editor for something that supports custom editor nodes.
3. I had to fix a couple of bugs in `vue-custom-elements`, so currently the dependency for that is a fork.  I'm gonna open PRs to see if I can get my fixes into the official version.

## Checklist


## Comments

This was a very fun project :smile:... I love web components!
